torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -694,161 +830,86 @@ def _step_microbatches(
-                bwd_mb_index := bwd_stage_mb_index[bwd_stage]
-            ) + 1
-
-            bwd_stage._configure_data_parallel_mode(
-                bwd_mb_index == self._n_microbatches - 1
-            )
-            logger.debug(
-                f"Rank {self.rank}: {step=}, {fwd_stage.stage_index=}, "  # noqa: G004
-                f"{bwd_stage.stage_index=}, {fwd_mb_index=}, {bwd_mb_index=}"
-            )
-            desc = f"1F1B {step}"
-            with record_function(desc):
-                ops.extend(fwd_stage.get_fwd_recv_ops())
-                ops.extend(bwd_stage.get_bwd_recv_ops())
-                if ops:
-                    work = _batch_p2p(ops, desc=desc)
-                    work.wait()
-                    ops.clear()
-
-                # Forward
-                output = fwd_stage.forward_one_chunk(arg_mbs[fwd_mb_index], kwarg_mbs[fwd_mb_index])  # type: ignore[index]
-                ops.extend(fwd_stage.get_fwd_send_ops())
-                self._maybe_compute_loss(fwd_stage, output, target_mbs, fwd_mb_index)
-
-                # Backward
-                loss = self._maybe_get_loss(bwd_stage, bwd_mb_index)
-                bwd_stage.backward_one_chunk(loss=loss)
-                ops.extend(bwd_stage.get_bwd_send_ops())
-
-        # Cooldown Phase (backward only)
-        for step in range(warmup_steps + fwd_bwd_steps, total_steps):
-            bwd_stage = self._stages[backward_stage_local_index(step)]
-            bwd_stage_mb_index[bwd_stage] = (
-                bwd_mb_index := bwd_stage_mb_index[bwd_stage]
-            ) + 1
-            bwd_stage._configure_data_parallel_mode(
-                bwd_mb_index == self._n_microbatches - 1
-            )
-
-            logger.debug(
-                f"Rank {self.rank}: {step=}, {bwd_stage.stage_index=}, {bwd_mb_index=}"  # noqa: G004
-            )
-            desc = f"Cooldown {step}"
-            with record_function(desc):
-                ops.extend(bwd_stage.get_bwd_recv_ops())
-                if ops:
-                    work = _batch_p2p(ops, desc=desc + " pre_bwd")
-                    work.wait()
-                    ops.clear()
-
-                loss = self._maybe_get_loss(bwd_stage, bwd_mb_index)
-                bwd_stage.backward_one_chunk(loss=loss)
-
-                ops.extend(bwd_stage.get_bwd_send_ops())
-                if ops:
-                    work = _batch_p2p(ops, desc=desc + " post_bwd")
-                    sends_to_wait.append(work)
-                    ops.clear()
-
-        # Make sure all sends are finished
-        for work in sends_to_wait:
-            work.wait()
-
+        # Based on the plan in Step 1 created in __init__:
+        # 2. Perform communication based on the pipeline_order
+        stage_index_to_stage: Dict[int, _PipelineStageBase] = {
+            stage.stage_index: stage for stage in self._stages
+        }
+        prev_rank: int = (self.rank - 1) % self.pp_group_size
```

Comment:
it looks like this schedule assumes there will be no skip connections.  If so, it needs to be asserted somewhere (or if it is maybe I missed it).

Reply:
Yeah the schedule assumes no skip connections. Checking / asserting for skip connections may be a bit more involved, I can do it in a follow up PR

Reply:
Fix needed in `ManualPipelineStage`

Reply:
follow up with an issue

Reply:
https://github.com/pytorch/pytorch/issues/127964